cclib/method/ddec.py

```diff
@@ -300,6 +327,324 @@ def calculate_refcharges(self):
+                    rightaverage = numpy.sum(grid_filter * rho_cond_sqrt) / num_grid_filter
+                    if leftaverage < 1e-20:
+                        self.tau[atomi][radiusi] = 0.0
+                    else:
+                        self.tau[atomi][radiusi] = numpy.divide(
+                            leftaverage,
+                            rightaverage,
+                            out=numpy.zeros_like(leftaverage),
+                            where=rightaverage != 0.0,
+                        )
+            # Make tau monotonic decreasing
+            self.tau[atomi] = numpy.maximum.accumulate(self.tau[atomi][::-1])[::-1]
+
+    def _psiai(self, ya, bigpsiAI, atomi):
+        # Re-evaluate cond_density
+        cond_density = ya + bigpsiAI * numpy.sqrt(ya)
+
+        # Monotonic Decrease Condition
+        cond_density = numpy.minimum.accumulate(cond_density)
+
+        # Re-evaluate psi_AI
+        psiAI = (
+            self._integrate_from_radial([cond_density], [atomi])
+            - self.data.atomnos[atomi]
+            + self.refcharges[-1][atomi]
+        )
+
+        return psiAI, cond_density
+
+    def _ya(self, proatom_density, atomi):
+        # Function that calculates Y_a^avg
+        # See Eq. 40-41 in doi: 10.1039/c6ra04656h
+        rho_ref = self._rho_ref
+        # Y_a^avg -- Equation 40 in doi: 10.1039/c6ra04656h
+        ya = numpy.zeros_like(proatom_density[atomi], dtype=float)
+        weights = self.chgdensity.data / rho_ref
+        summed_elec = 0
+        for radiusi in range(len(ya)):
+            grid_filter = self.closest_r_index[atomi] == radiusi
+            num_grid_filter = numpy.count_nonzero(grid_filter)
+            if num_grid_filter < 1:
+                ya[radiusi] = 0.0
+            else:
+                spherical_avg = numpy.sum(grid_filter * weights) / num_grid_filter
+                ya[radiusi] = proatom_density[atomi][radiusi] * spherical_avg
+
+        ya = numpy.maximum.accumulate(ya[::-1])[::-1]
+        ya = numpy.minimum.accumulate(ya)
+
+        # Normalize y_a (see module_DDEC6_valence_iterator.f08::284)
+        nelec = self._integrate_from_radial([ya], [atomi])
+        ya = ya * (self.data.atomnos[atomi] - self.refcharges[-1][atomi]) / nelec
+
+        return ya
+
+    def _integrate_from_radial(self, radial_density_list, atom_list):
+        # Function that reads in list of radial densities, projects it on Cartesian grid,
+        # and returns integrated value
+        grid = copy.deepcopy(self.chgdensity)
+        grid.data = numpy.zeros_like(grid.data)
+
+        xshape, yshape, zshape = self.chgdensity.data.shape
+        indices = ((x, y, z) for x in range(xshape) for y in range(yshape) for z in range(zshape))
+
+        for density, atomi in zip(radial_density_list, atom_list):
+            for x, y, z in indices:
+                grid.data[x][y][z] = (
+                    grid.data[x][y][z] + density[self.closest_r_index[atomi][x][y][z]]
```

Comment:
+= here

Reply:
Yes! I've tried to change any other similar cases as well. Thanks!